### PR TITLE
TRUNK-5146: Use Java streams in UserServiceImpl#checkPrivileges(User)

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/TRUNK-5146

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-

## Checklist: I completed these to help reviewers :)

- [ ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by sherrif10** [comment]
Am creating this new ticket because i heard merge conflicts during making new commts earleir, so am sorry for inconvenience , kindly review, in this ticket all the requirements as suggested earlier are inclusive. cc @ibacher , @mks-d  @dkayiwa .am hoping there is nothing missing as  wehad been discussing , thanks

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 1]
```diff
@@ -277,6 +281,49 @@ public void createUser_shouldNotAllowSystemIdEqualsUsernameWithLuhnCheckDigit()
...
+		try {
+			FieldUtils.getField(UserContext.class, "user", true).set(userContext, user);
+			internals.run();
+		} finally {
+			FieldUtils.getField(UserContext.class, "user", true).set(userContext, authenticatedUser);
+		}
+	}
+	@Test
+	public void createUser_shouldNotAllowAssigningSuperUserRoleIfCurrentUserDoesnotHaveAssignSystemDeveloperPrivilege() throws IllegalAccessException {
+
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java:325]: So the implementation here should:

1. Setup the logged in user as one who does not have the SUPER_USER role and who doesn't have the ASSIGN_SYSTEM_DEVELOPER privilege associated with any role they do have (as in the test I wrote above).

2. In side the `withCurrentUserAs()` function, tries to assign the SUPER_USER role to a new user.

3. We should verify that the message of the generated exception matches the `User.you.must.have.role` message.

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -395,32 +397,23 @@ public boolean isSecretAnswer(User u, String answer) {
...
 			throw new APIException("User.you.must.have.privilege", new Object[] { requiredPrivs.get(0) });
 		} else if (requiredPrivs.size() > 1) {
-			StringBuilder txt = new StringBuilder("You must have the following privileges in order to assign them: ");
-			for (String s : requiredPrivs) {
-				txt.append(s).append(", ");
-			}
-			throw new APIException(txt.substring(0, txt.length() - 2));
-		}
-	}
+			throw new APIException(String.format("User.you.must.have.privileges", String.join(", ", requiredPrivs)));
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java:407]: That means we are to catch this using try and catch block!, Any other option

File: api/src/test/java/org/openmrs/api/UserServiceTest.java (hunk 1 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java:332]: So am to write tests by the end of today but again may be formating the thrown exception into array. cc @ibacher , @mks-d ,

**Comment by sherrif10** [comment]
CI failure is being caused by a minor code , i hope to make fix it

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibacher) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java:407]: No. I just mean change this one line to: `throw new APIException(String.format("User.you.must.have.privileges", new Object[] { String.join(", ", requiredPrivs) }));`. This is to ensure we invoke the proper constructor for `APIException`.

File: api/src/test/java/org/openmrs/api/UserServiceTest.java (hunk 1 above)
Comment (tendomart) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java:331]: @sherrif10  , your PR is still noisy  as well. I also good practice is to write tests first , fail them then make them pass with your new changes.

**Comment by sherrif10** [comment]
Never mind it will look fine @tendomart , will all its neccessary tests because it  what am working on currently. thanks for your guide @tendomart

**Comment by sherrif10** [comment]
This has merge conficts ,and this may have been caused by latest merged branches, so we have to clone  a fresh master branch and close this, is there any other option @ibacher  @mks-d  @dkayiwa

**Comment by ibacher** [comment]
@sherrif10 I've fixed the merge commit for you. Generally to fix this kind of thing, I do something like this:

```
git fetch --prune origin
git rebase --onto origin/master
```

Then open the conflict files and find the:

```
<<<<<<< HEAD
...
=======
...
>>>>>>> master
```

Sections and figure out how best to resolve that. Either GitHub (which I used in this case) or [any decent GUI client](https://git-scm.com/downloads/guis/) can probably help you with the process (personally, I use [Sublime Merge](https://www.sublimemerge.com/), but have heard good things about [GitKraken](https://www.gitkraken.com/git-client))... [Tools in your IDE](https://www.vogella.com/tutorials/EclipseGit/article.html#solving-conflicts-created-by-merge-rebase-or-other-operations) might be able to help you too.

Unfortunately, [resolving conflicts](https://githowto.com/resolving_conflicts) is one of the more complex Git-related tasks, only because it always involves some judgement about what should stay and what should go.

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sherrif10) [reply api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java:407]: ok then thanks for clarification

**Comment by sherrif10** [comment]
Thanks @ibacher , actually i had tried but failed on the way,and found that to fix this , i have to have write access to the master, so i can go ahead and make new commits

**Comment by ibacher** [comment]
You really don't need write access to anything other than your branch to fix this. You just need to rebase your current branch on top of master.

**Comment by ibacher** [comment]
@sherrif10 I've fixed the conflict with this branch again. Can you kindly do a `git pull` on your current branch before continuing to code?

**Comment by sherrif10** [comment]
ok @ibacher thanks for good work here, i hope it was neccessary since we also faced the same challenge with reference TRUNK-5726. Seems i need to squash these commits again

**Comment by sherrif10** [comment]
@ibacher , am interactive rebase now, so should i go ahead and squash them

**Comment by ibacher** [comment]
Yes, you can do that

**Comment by ibacher** [comment]
I know this is a bit different from the recommended flow, but I don't usually focus on getting everything into a single commit until the code is ready to be merged... *sigh* another thing to bring up on talk.

**Comment by sherrif10** [comment]
Your are very true and i totally agree with you, i hope while squashing i will be rebasing this TRUNK as the HEAD

File: api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -544,6 +538,17 @@ private void checkPrivileges(Role role) {
...
+		Optional.ofNullable(role.getPrivileges())
+		.map(p -> p.stream().filter(pr -> !Context.hasPrivilege(pr.getPrivilege())).map(Privilege::getPrivilege)
+			.distinct().collect(Collectors.joining(", ")))
+		.ifPresent(missing -> {
+			if (StringUtils.isNotBlank(missing)) {
+				throw new APIException("Role.you.must.have.privileges", new Object[] { missing });
+			}
+		});
+    }
+>>>>>>> 119e3fa... Fix for TRUNK-5726
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java:551]: Can you fix this please?

File: api/src/main/java/org/openmrs/api/db/DelegatingFullTextSession.java [hunk 4]
```diff
@@ -134,10 +161,4 @@ public MassIndexer createIndexer(Class<?>... types) {
 	public FullTextSharedSessionBuilder sessionWithOptions() {
 		return delegate.sessionWithOptions();
 	}
-
-	@Override
-	public FullTextQuery createFullTextQuery(QueryDescriptor descriptor, Class<?>... entities) {
-		throw new UnsupportedOperationException();
-	}
-
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/db/DelegatingFullTextSession.java:142]: I'm not sure why this is included in this commit?

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 5]
```diff
@@ -277,6 +282,45 @@ public void createUser_shouldNotAllowSystemIdEqualsUsernameWithLuhnCheckDigit()
...
+			u.getPerson().setGender("M");
+			u.addRole(role);
+			// here we expect the exception to be thrown
+			userService.createUser(u, "Openmr5xy");
+		});
+	}
+	
+	@Test
+	public void createUser_shouldNotAllowAssigningSuperUserRoleIfCurrentUserDoesnotHaveAssignSystemDeveloperPrivilege() throws IllegalAccessException {
+
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java:322]: This test still needs an implementation

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 6]
```diff
@@ -1059,6 +1103,65 @@ public void saveRole_shouldThrowErrorIfRoleInheritsFromItself() {
...
+		expectedException.expectMessage("You must have the following privileges in order to assign them: another privilege");
+
+		withCurrentUserAs(currentUser, () -> {
+			Role newRole = new Role("another role");
+			newRole.addPrivilege(myFirstPrivilege);
+			newRole.addPrivilege(mySecondPrivilege);
+			userService.saveRole(newRole);
+		});
+	}
+
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java:1164]: Why is this in this PR?

File: api/src/test/java/org/openmrs/api/UserServiceTest.java [hunk 7]
```diff
@@ -1505,4 +1608,23 @@ public void changePasswordUsingActivationKey_shouldNotUpdatePasswordIfActivation
...
+	private void withCurrentUserAs(User user, Runnable internals) throws IllegalAccessException {
+		UserContext userContext = Context.getUserContext();
+		User authenticatedUser = userContext.getAuthenticatedUser();
+		try {
+			FieldUtils.getField(UserContext.class, "user", true).set(userContext, user);
+			internals.run();
+		} finally {
+			FieldUtils.getField(UserContext.class, "user", true).set(userContext, authenticatedUser);
+		}
+	}
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/api/UserServiceTest.java:1629]: This shouldn't be in this PR

**Comment by ibacher** [comment]
@sherrif10 Thanks for your work on this. It seems like this branch might've diverged pretty far from what's in master, since it seems to have the last two commits as part of the changes it needs? Also, there's still some Git conflict markers. Can you fix these things please?

File: api/src/main/java/org/openmrs/api/db/DelegatingFullTextSession.java (hunk 4 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/api/db/DelegatingFullTextSession.java:142]: this was among the new commits that were introduced, let me fix everything

File: api/src/test/java/org/openmrs/api/UserServiceTest.java (hunk 6 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java:1164]: This again came with the same PR but before this test was earlier better, so i guess it has no conflit

File: api/src/test/java/org/openmrs/api/UserServiceTest.java (hunk 7 above)
Comment (sherrif10) [reply api/src/test/java/org/openmrs/api/UserServiceTest.java:1629]: Have fixed them , everything here new came due to new commits after rebasing it, so they dont have any harm

**Comment by sherrif10** [comment]
Have force pushed, i hope it make more sense now, Other new commits have been merged actually into this because of rebasing the branch so currently they dont have any error, its running very well .

**Comment by sherrif10** [comment]
@ibacher  they are some current tests that am going to delete  which are making work bit rigid

**Comment by ibacher** [comment]
> @ibacher they are some current tests that am going to delete which are making work bit rigid

I wouldn't do that without a very good reason. Tests are there to ensure we don't break existing functionality. Why do you need to remove them and which tests do you propose removing?

**Comment by sherrif10** [comment]
This test 	@Test
	public void createUser_shouldNotAllowCreatingUserWithPrivilegeCurrentUserDoesNotHave() throws IllegalAccessException {
is breaking from here 	```
withCurrentUserAs(currentUser, () -> {
			// create a role to assign to the new user
			Role role = new Role();
			role.setRole(RoleConstants.AUTHENTICATED);
			role.addPrivilege(new Privilege("Custom Privilege"));// add a privilege to the role
			
```

**Comment by sherrif10** [comment]
I Closed this to create a new ticket with all tracking code that we need into this to make it more clean and updated

**Pull request closed without merging**